libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,500 @@
+//! # Buildpack output
+//!
+//! Use the [`BuildpackOutput`] to output structured text as a buildpack is executing.
+//!
+//! ```
+//! use libherokubuildpack::buildpack_output::BuildpackOutput;
+//!
+//! let mut output = BuildpackOutput::new(std::io::stdout())
+//!     .start("Heroku Ruby Buildpack")
+//!     .warning("No Gemfile.lock found");
+//!
+//! output = output
+//!     .section("Ruby version")
+//!     .finish();
+//!
+//! output.finish();
+//! ```
+//!
+use crate::buildpack_output::ansi_escape::{BOLD_PURPLE, CYAN, RED, YELLOW};
+use crate::buildpack_output::util::{prefix_first_rest_lines, prefix_lines, ParagraphInspectWrite};
+use crate::write::line_mapped;
+use std::fmt::Debug;
+use std::io::Write;
+use std::time::Instant;
+
+mod ansi_escape;
+mod duration_format;
+pub mod style;
+mod util;
+
+/// See the module docs for example usage.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug)]
+pub struct BuildpackOutput<T> {
+    pub(crate) started: Option<Instant>,
+    pub(crate) state: T,
+}
+
+/// Various states for [`BuildpackOutput`] to contain.
+///
+/// The [`BuildpackOutput`] struct acts as an output state machine. These structs
+/// are meant to represent those states.
+#[doc(hidden)]
+pub mod state {
```

Comment:
It doesn't seem like the `state` module needs to be public?

If I remove the `pub` here all of `clippy --all-targets`, `cargo test` and `cargo test -- --ignored` succeed. (And the rustdocs for the `build_output` module is exercising the code as a consumer would - rustdocs can't access private APIs etc.)

If this module does need to be public, could you add something that will fail if the `pub` is removed (eg rustdocs somewhere, an integration test, a brief usage in `examples/basics/` or similar etc), or at least a comment explaining why this is pub+doc-hidden?

```suggestion
mod state {
```

Reply:
The states must be exposed so people can use them in their type signatures. For example:

```
//! use libherokubuildpack::buildpack_output::{BuildpackOutput, state::{Section, Started}};
//! use std::io::Write;
//!
//! let mut output = BuildpackOutput::new(std::io::stdout())
//!     .start("Heroku Ruby Buildpack")
//!     .warning("No Gemfile.lock found");
//!
//!      let output = install_ruby(output.section("Ruby Version"));
//!      output.finish();
//!
//! fn install_ruby<W>(mut output: BuildpackOutput<Section<W>>) -> BuildpackOutput<Started<W>>
//! where
//!    W: Write + Send + Sync + 'static
//! {
//!     output = output.step("Installing Ruby");
//!     // ...
//!    output.finish()
//! }
```

Changing that module to a different visibility makes it error:

```
5  | use libherokubuildpack::buildpack_output::{BuildpackOutput, state::{Section, Started}};
   |                                                             ^^^^^ private module
```

I'll add some more failing doc tests that exercise each state